contracts/chain-adapters/Arbitrum_CustomGasToken_Adapter.sol

```diff
@@ -0,0 +1,291 @@
+            message // data ABI encoded data of L2 message
+        );
+        emit MessageRelayed(target, message);
+    }
+
+    /**
+     * @notice Bridge tokens to Arbitrum.
+     * @notice This contract must hold at least getL1CallValue() amount of ETH or custom gas token
+     * to send a message via the Inbox successfully, or the message will get stuck.
+     * @param l1Token L1 token to deposit.
+     * @param l2Token L2 token to receive.
+     * @param amount Amount of L1 tokens to deposit and L2 tokens to receive.
+     * @param to Bridge recipient.
+     */
+    function relayTokens(
+        address l1Token,
+        address l2Token, // l2Token is unused for Arbitrum.
+        uint256 amount,
+        address to
+    ) external payable override {
+        // Check if this token is USDC, which requires a custom bridge via CCTP.
+        if (_isCCTPEnabled() && l1Token == address(usdcToken)) {
+            _transferUsdc(to, amount);
+        }
+        // If not, we can use the Arbitrum gateway
+        else {
+            address erc20Gateway = L1_ERC20_GATEWAY_ROUTER.getGateway(l1Token);
+
+            // If custom gas token, call special functions that handle paying with custom gas tokens.
+            uint256 requiredL1TokenTotalFeeAmount = _pullCustomGas(RELAY_MESSAGE_L2_GAS_LIMIT);
+
+            // Must use Inbox to bridge custom gas token.
+            // Source: https://github.com/OffchainLabs/token-bridge-contracts/blob/5bdf33259d2d9ae52ddc69bc5a9cbc558c4c40c7/contracts/tokenbridge/ethereum/gateway/L1OrbitERC20Gateway.sol#L33
+            if (l1Token == address(CUSTOM_GAS_TOKEN)) {
+                uint256 amountToBridge = amount + requiredL1TokenTotalFeeAmount;
+                CUSTOM_GAS_TOKEN.safeIncreaseAllowance(address(L1_INBOX), amountToBridge);
+                L1_INBOX.createRetryableTicket(
+                    to, // destAddr destination L2 contract address
+                    L2_CALL_VALUE, // l2CallValue call value for retryable L2 message
+                    L2_MAX_SUBMISSION_COST, // maxSubmissionCost Max gas deducted from user's L2 balance to cover base fee
+                    L2_REFUND_L2_ADDRESS, // excessFeeRefundAddress maxgas * gasprice - execution cost gets credited here on L2
+                    L2_REFUND_L2_ADDRESS, // callValueRefundAddress l2Callvalue gets credited here on L2 if retryable txn times out or gets cancelled
+                    RELAY_MESSAGE_L2_GAS_LIMIT, // maxGas Max gas deducted from user's L2 balance to cover L2 execution
+                    L2_GAS_PRICE, // gasPriceBid price bid for L2 execution
+                    amountToBridge, // tokenTotalFeeAmount amount of fees to be deposited in native token.
+                    "0x" // data ABI encoded data of L2 message
+                );
+            } else {
+                CUSTOM_GAS_TOKEN.safeIncreaseAllowance(erc20Gateway, amount);
```

Comment:
nit: should this be `requiredL1TokenTotalFeeAmount` instead of `amount`?

Reply:
Yes great catch. I think I'm also missing an increaseAllowance of the l1Token in this else{} branch: 9674ed0f9fbecb6a1fda87e9d5081a91eb47042c